### ISSUE TITLE
Bug: PIA TLS error + close of closed channel

### ISSUE THREAD

**Issue by ACNiC03** [qdm12/gluetun#999]
### Host OS

Debian 11

### CPU arch

x86_64

### VPN service provider

Private Internet Access

### What are you using to run the container

Portainer

### What is the version of Gluetun

Running version latest built on 2022-05-07T07:18:37.602Z (commit e32d251)

### What's the problem 🤔

I cannot seem to connect to the PIA servers. It gives an tls error issue. 

I am trying to figure out what is wrong with my configuration... I started basic with the default compose configuration. it showed errors about the known tun module, it looked like it got fixed when i addedd privileged: true to the compose file.

After that error i got the openvpn errors, those looked fixed when i added "VPN_ENDPOINT_PORT=1198". but now it gives this error below. so what am i doing wrong??

### Share your logs

```log
Running version latest built on 2022-05-07T07:18:37.602Z (commit e32d251)
2022-05-21T20:51:15+02:00 INFO [openvpn] TCP/UDP: Preserving recently used remote address: [AF_INET]212.102.35.22:1198
2022-05-21T20:51:15+02:00 INFO [openvpn] UDP link local: (not bound)
2022-05-21T20:51:15+02:00 INFO [openvpn] UDP link remote: [AF_INET]212.102.35.22:1198
2022-05-21T20:51:15+02:00 INFO [openvpn] VERIFY ERROR: depth=1, error=self signed certificate in certificate chain: C=US, ST=CA, L=LosAngeles, O=Private Internet Access, OU=Private Internet Access, CN=Private Internet Access, name=Private Internet Access, emailAddress=secure@privateinternetaccess.com, serial=11996199170155461251
2022-05-21T20:51:15+02:00 INFO [openvpn] OpenSSL: error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed
2022-05-21T20:51:15+02:00 INFO [openvpn] TLS_ERROR: BIO read tls_read_plaintext error
2022-05-21T20:51:15+02:00 INFO [openvpn] TLS Error: TLS object -> incoming plaintext read error
2022-05-21T20:51:15+02:00 INFO [openvpn] TLS Error: TLS handshake failed
2022-05-21T20:51:15+02:00 INFO [openvpn] SIGTERM received, sending exit notification to peer
2022-05-21T20:51:15+02:00 INFO [openvpn] SIGTERM[soft,tls-error] received, process exiting
goroutine 94 [running]:
panic: close of closed channel
github.com/qdm12/gluetun/internal/vpn.(*Loop).Run(0xc0005ec280, {0xd516f8, 0xc00059f480}, 0xc00002fe00)
created by main._main
	github.com/qdm12/gluetun/internal/vpn/run.go:99 +0xcb9
	./main.go:386 +0x2e85
```

### Share your configuration

```yml
---
version: "3"
services:
  gluetun:
    image: qmcgaw/gluetun
    container_name: gluetun
    cap_add:
      - NET_ADMIN
    privileged: true
    devices:
      - /dev/net/tun:/dev/net/tun
    ports:
      - 9696:9696 # Prowlarr
      - 8080:8080 # Qbit WebUI
      - 6881:6881 # Qbit
      - 6881:6881/udp # Qbit
    volumes:
      - /data/Gluetun:/gluetun
    environment:
      - UID=0
      - GID=0
      - VPN_SERVICE_PROVIDER=private internet access
      - OPENVPN_USER=...
      - OPENVPN_PASSWORD=...
      #- VPN_ENDPOINT_IP=176.34.241.253
      - VPN_ENDPOINT_PORT=1198
      - SERVER_REGIONS=Netherlands
      - TZ=Europe/Amsterdam
      - UPDATER_PERIOD=24h
      #- SERVER_HOSTNAME=nl-amsterdam.privacy.network
    #dns:
     # - 192.168.1.1 # Pfsense
     # - 1.1.1.1 # Cloudflare
    restart: always
```

**Comment by ACNiC03** [comment]
For everyone wondering why it is not working... DONT USE SPECIAL CHARACTERS IN YOUR PASSWORD

**Comment by qdm12** [comment]
Hey there, sorry for the delay answering.

> VERIFY ERROR: depth=1, error=self signed certificate in certificate chain

This might be due to #958? Does it fail if you repull the latest image (or v3.29.0)? I'm not sure your credentials would influence this 🤔

> panic: close of closed channel

That's a programming error on my side, I'll investigate it, please leave this issue opened 😉

**Comment by ACNiC03** [comment]
@qdm12 I have changed my password to a password without special characters. It seemed to have it fixed. It workes flawlessly now. Maybe it is a good idea to have a note to use a password without special characters? After changing my password i also deleted the lines:
privileged: true
devices:
   - /dev/net/tun:/dev/net/tun

Even after that it still worked. So it looks really like that the special characters in my password were the errorcreators.

**Comment by ACNiC03** [comment]
I have a new error after Gluetun worked great for a few days. The error that is shown now is:
**INFO [openvpn] read UDP [ECONNREFUSED]: Connection refused (code=111)**

I have changed nothing on the compose file. Any idea what the problem could be? @qdm12

**Comment by tjxme** [comment]
i took away my special characters, still shows TLS error.
|       ├── Password: [set]
|       ├── Private Internet Access encryption preset: strong
|       ├── Tunnel IPv6: no
|       ├── Network interface: tun0
|       ├── Run OpenVPN as: root
|       └── Verbosity level: 1
├── DNS settings:
|   ├── DNS server address to use: 127.0.0.1
|   ├── Keep existing nameserver(s): no
|   └── DNS over TLS settings:
|       ├── Enabled: yes
|       ├── Update period: every 24h0m0s
|       ├── Unbound settings:
|       |   ├── Authoritative servers:
|       |   |   └── cloudflare
|       |   ├── Caching: yes
|       |   ├── IPv6: no
|       |   ├── Verbosity level: 1
|       |   ├── Verbosity details level: 0
|       |   ├── Validation log level: 0
|       |   ├── System user: root
|       |   └── Allowed networks:
|       |       ├── 0.0.0.0/0
|       |       └── ::/0
|       └── DNS filtering settings:
|           ├── Block malicious: yes
|           ├── Block ads: no
|           ├── Block surveillance: no
|           └── Blocked IP networks:
|               ├── 127.0.0.1/8
|               ├── 10.0.0.0/8
|               ├── 172.16.0.0/12
|               ├── 192.168.0.0/16
|               ├── 169.254.0.0/16
|               ├── ::1/128
|               ├── fc00::/7
|               ├── fe80::/10
|               ├── ::ffff:7f00:1/104
|               ├── ::ffff:a00:0/104
|               ├── ::ffff:a9fe:0/112
|               ├── ::ffff:ac10:0/108
|               └── ::ffff:c0a8:0/112
├── Firewall settings:
|   └── Enabled: yes
├── Log settings:
|   └── Log level: INFO
├── Health settings:
|   ├── Server listening address: 127.0.0.1:9999
|   ├── Target address: cloudflare.com:443
|   └── VPN wait durations:
|       ├── Initial duration: 6s
|       └── Additional duration: 5s
├── Shadowsocks server settings:
|   └── Enabled: no
├── HTTP proxy settings:
|   └── Enabled: no
├── Control server settings:
|   ├── Listening address: :8000
|   └── Logging: yes
├── OS Alpine settings:
|   ├── Process UID: 1026
|   ├── Process GID: 100
|   └── Timezone: est
├── Public IP settings:
|   ├── Fetching: every 12h0m0s
|   └── IP file path: /tmp/gluetun/ip
└── Version settings:
    └── Enabled: yes
2022-06-05T22:31:50Z INFO [routing] default route found: interface eth0, gateway 172.24.0.1 and assigned IP 172.24.0.2
2022-06-05T22:31:50Z INFO [routing] adding route for 0.0.0.0/0
2022-06-05T22:31:50Z INFO [firewall] setting allowed subnets...
2022-06-05T22:31:50Z INFO [routing] default route found: interface eth0, gateway 172.24.0.1 and assigned IP 172.24.0.2
2022-06-05T22:31:50Z INFO TUN device is not available: open /dev/net/tun: no such file or directory; creating it...
2022-06-05T22:31:50Z INFO [pprof] http server listening on [::]:6060
2022-06-05T22:31:50Z INFO [dns over tls] using plaintext DNS at address 1.1.1.1
2022-06-05T22:31:50Z INFO [http server] http server listening on [::]:8000
2022-06-05T22:31:50Z INFO [healthcheck] listening on 127.0.0.1:9999
2022-06-05T22:31:50Z INFO [firewall] allowing VPN connection...
2022-06-05T22:31:50Z INFO [openvpn] OpenVPN 2.5.6 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Mar 24 2022
2022-06-05T22:31:50Z INFO [openvpn] library versions: OpenSSL 1.1.1o  3 May 2022, LZO 2.10
2022-06-05T22:31:50Z INFO [openvpn] CRL: loaded 1 CRLs from file -----BEGIN X509 CRL-----
2022-06-05T22:31:50Z INFO [openvpn] MIIDWDCCAUAwDQYJKoZIhvcNAQENBQAwgegxCzAJBgNVBAYTAlVTMQswCQYDVQQIEwJDQTETMBEGA1UEBxMKTG9zQW5nZWxlczEgMB4GA1UEChMXUHJpdmF0ZSBJbnRlcm5ldCBBY2Nlc3MxIDAeBgNVBAsTF1ByaXZhdGUgSW50ZXJuZXQgQWNjZXNzMSAwHgYDVQQDExdQcml2YXRlIEludGVybmV0IEFjY2VzczEgMB4GA1UEKRMXUHJpdmF0ZSBJbnRlcm5ldCBBY2Nlc3MxLzAtBgkqhkiG9w0BCQEWIHNlY3VyZUBwcml2YXRlaW50ZXJuZXRhY2Nlc3MuY29tFw0xNjA3MDgxOTAwNDZaFw0zNjA3MDMxOTAwNDZaMCYwEQIBARcMMTYwNzA4MTkwMDQ2MBECAQYXDDE2MDcwODE5MDA0NjANBgkqhkiG9w0BAQ0FAAOCAgEAppFfEpGsasjB1QgJcosGpzbf2kfRhM84o2TlqY1ua+Gi5TMdKydA3LJcNTjlI9a0TYAJfeRX5IkpoglSUuHuJgXhP3nEvX10mjXDpcu/YvM8TdE5JV2+EGqZ80kFtBeOq94WcpiVKFTR4fO+VkOK9zwspFfb1cNs9rHvgJ1QMkRUF8PpLN6AkntHY0+6DnigtSaKqldqjKTDTv2OeH3nPoh80SGrt0oCOmYKfWTJGpggMGKvIdvU3vH9+EuILZKKIskt+1dwdfA5Bkz1GLmiQG7+9ZZBQUjBG9Dos4hfX/rwJ3eU8oUIm4WoTz9rb71SOEuUUjP5NPy9HNx2vx+cVvLsTF4ZDZaUztW9o9JmIURDtbeyqxuHN3prlPWB6aj73IIm2dsDQvs3XXwRIxs8NwLbJ6CyEuvEOVCskdM8rdADWx1J0lRNlOJ0Z8ieLLEmYAA834VN1SboB6wJIAPxQU3rcBhXqO9y8aa2oRMg8NxZ5gr+PnKVMqag1x0IxbIgLxtkXQvxXxQHEMSODzvcOfK/nBRBsqTj30P+R87sU8titOoxNeRnBDRNhdEy/QGAqGh62ShPpQUCJdnKRiRTjnil9hMQHevoSuFKeEMO30FQL7BZyo37GFU+q1WPCplVZgCP9hC8Rn5K2+f6KLFo5bhtowSmu+GY1yZtg+RTtsA=
2022-06-05T22:31:50Z INFO [openvpn] -----END X509 CRL-----
2022-06-05T22:31:50Z INFO [openvpn] TCP/UDP: Preserving recently used remote address: [AF_INET]154.3.40.43:1197
2022-06-05T22:31:50Z INFO [openvpn] UDP link local: (not bound)
2022-06-05T22:31:50Z INFO [openvpn] UDP link remote: [AF_INET]154.3.40.43:1197
2022-06-05T22:31:50Z INFO [openvpn] VERIFY ERROR: depth=0, error=certificate has expired: C=US, ST=CA, L=LosAngeles, O=Private Internet Access, OU=Private Internet Access, CN=toronto415, name=toronto415, serial=95429892483
2022-06-05T22:31:50Z INFO [openvpn] OpenSSL: error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed
2022-06-05T22:31:50Z INFO [openvpn] TLS_ERROR: BIO read tls_read_plaintext error
2022-06-05T22:31:50Z INFO [openvpn] TLS Error: TLS object -> incoming plaintext read error
2022-06-05T22:31:50Z INFO [openvpn] TLS Error: TLS handshake failed
2022-06-05T22:31:50Z INFO [openvpn] SIGTERM received, sending exit notification to peer
2022-06-05T22:31:50Z INFO [openvpn] SIGTERM[soft,tls-error] received, process exiting
panic: close of closed channel

goroutine 96 [running]:
github.com/qdm12/gluetun/internal/vpn.(*Loop).Run(0xc0002ba140, {0xd53378, 0xc00086f580}, 0xc00002f140)
	github.com/qdm12/gluetun/internal/vpn/run.go:99 +0xbd9
created by main._main
	./main.go:386 +0x2d05

**Comment by ACNiC03** [comment]
@tjxme did you update the hostnames/IP's? A few days ago those were changed from PIA...

**Comment by tjxme** [comment]
> @tjxme did you update the hostnames/IP's? A few days ago those were changed from PIA...

I didn't specify any host names besides host_region=ca Toronto. 

Also, I changed the vpn name to full name as indicated on wiki.
Still not working

**Comment by ACNiC03** [comment]
> > @tjxme did you update the hostnames/IP's? A few days ago those were changed from PIA...
> 
> I didn't specify any host names besides host_region=ca Toronto.
> 
> Also, I changed the vpn name to full name as indicated on wiki. Still not working

Try this on your host:
docker run -it --rm -v /<yourpath>:/gluetun qmcgaw/gluetun update -enduser -providers 'private internet access'

**Comment by tjxme** [comment]
> > > @tjxme did you update the hostnames/IP's? A few days ago those were changed from PIA...
> > 
> > 
> > I didn't specify any host names besides host_region=ca Toronto.
> > Also, I changed the vpn name to full name as indicated on wiki. Still not working
> 
> Try this on your host: docker run -it --rm -v /_yourpath_:/gluetun qmcgaw/gluetun update -enduser -providers 'private internet access'

im using compose, can you offer me a version of that?

**Comment by ACNiC03** [comment]
> im using compose, can you offer me a version of that please? thanks a lot

Doesnt compose still use docker to run containers? I think it wouldnt make any difference, i am not so much educated on docker-compose other then creating a yml file to run containers...

So if you want a specific fix with compose i am the wrong person for that...

**Comment by qdm12** [comment]
@tjxme I added a yml example to the [wiki page](https://github.com/qdm12/gluetun/wiki/Updating-servers) but you're probably better off just running the docker command which is simpler. Also built-in PIA servers data got updated today in e95c94294fe37e67ad5307e6696b111826919989 so you could just `docker pull qmcgaw/gluetun` if you prefer.
I'll hide your comments since they're a bit off-topic, I would like to keep this issue focused on the panic error now.

**Comment by qdm12** [comment]
The panic should be fixed with b71c8e58f45b98761cb84e662c6245a43d4d5798

I'll rework most of the 'loop async logic' as mentioned in https://github.com/qdm12/gluetun/blob/85e9d7d522ea98f54a0f509e789352344e3999e7/maintenance.md#L4 later to have more robust logic. Closing this for now!
